AnyText/AnyText.Core/Parser.SemanticToken.cs

```diff
@@ -0,0 +1,129 @@
+        ///     If null, all tokens to the end are considered.
+        /// </param>
+        /// <returns>
+        ///     A list of semantic tokens represented as unsigned integers,
+        ///     including delta-encoded line and character positions, length, token type, and modifiers.
+        /// </returns>
+        public IEnumerable<uint> GetSemanticElementsFromRoot(ParsePosition? start = null, ParsePosition? end = null)
+        {
+            var semanticTokens = new List<uint>();
+            
+            var rootApplication = Context.RootRuleApplication;
+            uint previousLine = 0;
+            uint previousStartChar = 0;
+            rootApplication.IterateLiterals(literalRuleApp =>
+            {  
+                if (start.HasValue && end.HasValue && !IsTokenInRange(literalRuleApp.CurrentPosition, literalRuleApp.ExaminedTo, start.Value, end.Value))
+                {
+                    return;
+                }
+                
+                var line = (uint)literalRuleApp.CurrentPosition.Line;
+                var startChar = (uint)literalRuleApp.CurrentPosition.Col;
+                var modifiers = GetTokenModifierIndexFromHierarchy(literalRuleApp) ?? 0;
+                var tokenType = GetTokenTypeIndexFromHierarchy(literalRuleApp) ?? 0;
+
+                //split literal into lines for multiline literals to assign tokens (comments)
+                var lines = literalRuleApp.Literal.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None);
+                var currentLine = line;
+                foreach (var lineText in lines)
+                {
+                    if (string.IsNullOrWhiteSpace(lineText))
+                    {
+                        currentLine++;
+                        continue;
+                    }
+
+                    var currentLength = (uint)lineText.Length;
+
+                    var deltaLine = currentLine - previousLine;
+                    var deltaStartChar = deltaLine == 0 ? startChar - previousStartChar : startChar;
+
+                    semanticTokens.Add(deltaLine);
+                    semanticTokens.Add(deltaStartChar);
+                    semanticTokens.Add(currentLength);
+                    semanticTokens.Add(tokenType);
+                    semanticTokens.Add(modifiers);
+
+                    previousLine = currentLine;
+                    previousStartChar = startChar;
+
+                    startChar = 0;
+                    currentLine++;
+                }
+            });
+
+
+            return semanticTokens;
+        }
+        /// <summary>
+        ///     Checks whether a token's range is within a specified range.
+        /// </summary>
+        private bool IsTokenInRange(ParsePosition start, ParsePositionDelta examinedTo, ParsePosition rangeStart, ParsePosition rangeEnd)
+        {
+            uint tokenEndLine = (uint) (start.Line + examinedTo.Line);
```

Comment:
Ich würde hier nicht nach uint konvertieren. Wir kommen schon nicht in den Bereich von >2 Mrd. Zeilen